keras_cv/layers/preprocessing/mosaic.py

```diff
@@ -0,0 +1,293 @@
+        x1 = (input_width - mosaic_centers[index][0]) / (input_width * 2 - 1)
+        y1 = (input_height - mosaic_centers[index][1]) / (input_height * 2 - 1)
+        x2 = x1 + (input_width) / (input_width * 2 - 1)
+        y2 = y1 + (input_height) / (input_height * 2 - 1)
+
+        # helps avoid retracing caused by slicing, inspired by RRC implementation
+        output = tf.image.crop_and_resize(
+            tf.expand_dims(output, axis=0),
+            [[y1, x1, y2, x2]],
+            [0],
+            [input_height, input_width],
+        )
+        return tf.squeeze(output)
+
+    def _update_label(self, images, labels, permutation_order, mosaic_centers, index):
+        # updates labels for one output mosaic
+        input_height, input_width, _ = images.shape[1:]
+        labels_for_mosaic = tf.gather(labels, permutation_order[index])
+        center_x = mosaic_centers[index][0]
+        center_y = mosaic_centers[index][1]
+
+        area = input_height * input_width
+
+        # labels are in the same ratio as the area of the images
+        top_left_ratio = (center_x * center_y) / area
+        top_right_ratio = ((input_width - center_x) * center_y) / area
+        bottom_left_ratio = (center_x * (input_height - center_y)) / area
+        bottom_right_ratio = (
+            (input_width - center_x) * (input_height - center_y)
+        ) / area
+
+        label = (
+            labels_for_mosaic[0] * top_left_ratio
+            + labels_for_mosaic[1] * top_right_ratio
+            + labels_for_mosaic[2] * bottom_left_ratio
+            + labels_for_mosaic[3] * bottom_right_ratio
+        )
+        return label
+
+    def _update_bounding_box(
+        self, images, bounding_boxes, permutation_order, translate_x, translate_y, index
+    ):
+        # updates bboxes for one output mosaic
+        bounding_boxes = bounding_box.convert_format(
+            bounding_boxes,
+            source=self.bounding_box_format,
+            target="xyxy",
+            images=images,
+        )
+
+        boxes_for_mosaic = tf.gather(bounding_boxes, permutation_order[index])
+        boxes_for_mosaic, rest = tf.split(
+            boxes_for_mosaic, [4, bounding_boxes.shape[-1] - 4], axis=-1
+        )
+
+        # stacking translate values such that the shape is (4, 1, 4) or (num_images, broadcast dim, coordinates)
+        translate_values = tf.stack(
+            [
+                translate_x[index],
+                translate_y[index],
+                translate_x[index],
+                translate_y[index],
+            ],
+            axis=-1,
+        )
+        translate_values = tf.expand_dims(translate_values, axis=1)
+
+        # translating boxes
+        boxes_for_mosaic = boxes_for_mosaic + translate_values
+        boxes_for_mosaic = tf.concat([boxes_for_mosaic, rest], axis=-1)
+        boxes_for_mosaic = tf.reshape(boxes_for_mosaic, [-1, bounding_boxes.shape[-1]])
+
+        boxes_for_mosaic = bounding_box.clip_to_image(
```

Comment:
basically bounding boxes that are shifted out of bounds will need to have their class_id set to -1.

we should update clip_to_image() to also filter out boxes that are fully outside of the image - and assign them -1 class ids.